### PR TITLE
Extract JsonWalker

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Follow up of #6187 and https://github.com/SonarSource/sonar-dotnet/pull/6187#discussion_r992050938

`Squash S206-appsettings` doesn't need a review

## Original discussion:

@pavel-mikula-sonarsource 
>Extract as reusable walker?

@martin-strecker-sonarsource 
>Sounds like a good idea to me. You could provide `IEnumerable<JsonNode> Descendants()` and `IEnumerable<KeyValuePair<string?, object>> DescendantValues()` extension methods for `JsonNode` on top of it.
>
>It's also easier to test the visitor for things like nested lists and the like.

@pavel-mikula-sonarsource 
>The walker is not as easy at it seemed. I'll leave it like this for now and we can discuss it here and do it in another PR.
The key problem is, that I need to visit `"value"`, but I do not want to visit `"value"` inside object `"key": "value"`. And node doesn't know it's parent.
I don't have a good idea for this stop mechanism without going into recursive descent and enqueuing in the `base.VisitObject`.
>
>How would `Descendants' and `DescendantValues` actually work together?

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+        public void VisitsAllNodes()
+        {
+            const string json = @"
+{
+    ""OuterKey"": ""OuterValue"",
+    ""OuterBool"": true,
+    ""NestedArray"": [
+        ""Array1"",
+        ""Array2"",
+        ""Array3""
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs:38]: It's not nested:
```suggestion
    ""NestedArray"": [
        ""Array1"",
        [""Array2-1"", ""Array2-2"", { InnerKey: ""Array2-3"" }, null],
        ""Array3""
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    [TestClass]
+    public class JsonWalkerTest
+    {
+        [TestMethod]
+        public void VisitsAllNodes()
+        {
+            const string json = @"
+{
+    ""OuterKey"": ""OuterValue"",
+    ""OuterBool"": true,
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs:34]: One null object should be added:
```suggestion
    ""OuterBool"": true,
    Nothing: null,
```

File: analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+using SonarAnalyzer.Json;
+
+namespace SonarAnalyzer.UnitTest.Common
+{
+    [TestClass]
+    public class JsonWalkerTest
+    {
+        [TestMethod]
+        public void VisitsAllNodes()
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs:29]: Add a parameterized test with empty json:
* `[]`
* `{}`
* `null`
* `""`
* `0`
* `true`
* `false`

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
Looks good, but some more tests are needed.

File: analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs:29]: Atomic values are not supported for parsing, I've added only `[]` and `{}`

Comment (martin-strecker-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs]: > Atomic values are not supported for parsing

Those are valid json but as long as we don't have a problem with it, that's fine. Is it worth fixing it?

File: analyzers/src/SonarAnalyzer.Common/Json/JsonWalker.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.Json.Parsing;
+
+namespace SonarAnalyzer.Json
+{
+    public abstract class JsonWalker
+    {
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Json/JsonWalker.cs:26]: Fix the code smell https://sonarcloud.io/project/issues?pullRequest=6201&issues=AYPLr8WRbucSp08jW4_H&open=AYPLr8WRbucSp08jW4_H&id=sonaranalyzer-dotnet

```suggestion
    public class JsonWalker
    {
        protected JsonWalker() { }

```

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
1 code smell to fix

File: analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/Json/JsonWalkerTest.cs]: No. There's no point in analyzing `true` or `0` file during analysis, there's nothing to learn about the project configuration there

File: analyzers/src/SonarAnalyzer.Common/Json/JsonWalker.cs (hunk 4 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Json/JsonWalker.cs]: I have very mixed feeling about this suggestion. It IMO just moves the "inherit me" (aka `MustInherit` in VB) information somewhere deep in the code, making it less readable.

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Json/JsonWalker.cs]: Same here. And the description of the rule has a flaw too:
https://github.com/SonarSource/rspec/pull/1341

**Review by martin-strecker-sonarsource — APPROVED** [review]
LGTM
